pkg/controller/replication/replication_controller.go

```diff
@@ -632,5 +634,66 @@ func (rm *ReplicationManager) syncReplicationController(key string) error {
 		updatedRC.Status.AvailableReplicas != *(updatedRC.Spec.Replicas) {
 		rm.enqueueControllerAfter(updatedRC, time.Duration(updatedRC.Spec.MinReadySeconds)*time.Second)
 	}
+
+	pods, err := rm.podLister.Pods(rc.Namespace).List(labels.Set(rc.Spec.Selector).AsSelectorPreValidated())
+	if err != nil {
+		return err
+	}
+	// Find rejected pods and delete them.
+	for _, pod := range pods {
+		if controller.IsPodRejected(pod) {
+			// Ignore deletion error here.
+			rm.podControl.DeletePod(pod.Namespace, pod.Name, rc)
+		}
+	}
+
+	boolPtr := func(b bool) *bool { return &b }
+	controllerRef := &metav1.OwnerReference{
+		APIVersion:         controllerKind.GroupVersion().String(),
+		Kind:               controllerKind.Kind,
+		Name:               rc.Name,
+		UID:                rc.UID,
+		BlockOwnerDeletion: boolPtr(true),
+		Controller:         boolPtr(true),
+	}
+	if pod, err := controller.GetPodFromTemplate(rc.Spec.Template, rc, controllerRef); err == nil {
+		if err = rm.validateNodeConflict(pod); err != nil {
+			return err
+		}
+	}
+
 	return manageReplicasErr
 }
+
+func (rm *ReplicationManager) validateNodeConflict(pod *v1.Pod) error {
+	if len(pod.Spec.NodeName) == 0 {
+		return nil
+	}
+
+	node, err := rm.kubeClient.Core().Nodes().Get(pod.Spec.NodeName, metav1.GetOptions{})
```

Comment:
I agree with the principle of generic mechanism. But unfortunately, the RC is not creating the same pods.

```
                "selfLink": "/api/v1/namespaces/default/pods/rc-nginx-single-42s86",
                "uid": "3eeae81c-30bf-11e7-8b20-fa163e09f30e"
```

```
                "selfLink": "/api/v1/namespaces/default/pods/rc-nginx-single-xvd97",
                "uid": "3dd8927c-30bf-11e7-8b20-fa163e09f30e"
```

After a pod being rejected by kubelet, the RC didn't care about that pod and do nothing for it. The RC just finds that it doesn't have enough replicas and then try to create new pods. So in this PR I delete the rejected pods and check if MatchNodeSelector fails.